### PR TITLE
[BUGFIX] optional dataframe on datasources

### PR DESCRIPTION
**Pull request by tyler-hoffman**
- [x] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [x] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [x] Code is linted

    ```
    black .

    ruff . --fix
    ```
- [x] Appropriate tests and docs have been updated

For more details, see our [Contribution Checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist), [Coding style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style), and [Documentation style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/docs_style).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

### CODE REVIEW

**Comment by tyler-hoffman** [comment]
> @cla-bot check

@cla-bot check

File: great_expectations/datasource/fluent/pandas_datasource.py [hunk 1]
```diff
@@ -581,7 +592,7 @@ def add_dataframe_asset(
         Returns:
             The DataFameAsset that has been added to this datasource.
         """
-        asset = DataFrameAsset(
+        asset: DataFrameAsset = DataFrameAsset(
```
Comment (NathanFarmer) [inline great_expectations/datasource/fluent/pandas_datasource.py:595]: Did the type checker require this change? Typically we don't do this for "self-evident" types, but if `mypy` wants it you can leave it like this.

File: tests/datasource/fluent/conftest.py [hunk 2]
```diff
@@ -267,6 +267,34 @@ def _put_db_datasources_callback(
...
+            200, headers=_DEFAULT_HEADERS, body=json.dumps(payload)
+        )
+    else:
+        errors = ErrorPayload(
+            errors=[{"code": "mock 404", "detail": None, "source": None}]
+        )
+        result = _CallbackResult(404, headers=_DEFAULT_HEADERS, body=json.dumps(errors))
+
+    logger.info(f"Response {result.status}")
+    return result
```
Comment (NathanFarmer) [inline tests/datasource/fluent/conftest.py:295]: ❤️

**Review by NathanFarmer — APPROVED** [review]

File: great_expectations/datasource/fluent/pandas_datasource.py (hunk 1 above)
Comment (tyler-hoffman) [reply great_expectations/datasource/fluent/pandas_datasource.py:595]: Yeah, I didn't want to put that in, but for whatever reason, the type checker required it 🤷

**Pull request merged**